zerver/lib/user_groups.py

```diff
@@ -365,13 +367,12 @@ def get_user_group_member_ids(
 
 def get_subgroup_ids(user_group: UserGroup, *, direct_subgroup_only: bool = False) -> List[int]:
     if direct_subgroup_only:
-        subgroup_ids = user_group.direct_subgroups.all().values_list("id", flat=True)
-    else:
-        subgroup_ids = (
-            get_recursive_subgroups(user_group)
-            .exclude(id=user_group.id)
-            .values_list("id", flat=True)
-        )
+        direct_subgroup_ids = user_group.direct_subgroups.all().values_list("id", flat=True)
+        return list(direct_subgroup_ids)
+
+    subgroup_ids = (
+        get_recursive_subgroups(user_group).exclude(id=user_group.id).values_list("id", flat=True)
```

Comment:
Needed this change to keep mypy happy.

`subgroup_ids` for the `direct_subgroup_only` case is of type `Queryset[NamedUserGroup, int]` because `direct_subgroups` references `NamedUserGroup` but in the else condition, `subgroup_ids` is of type `Queryset[UserGroup, int]` because the `get_recursive_subgroups` call returns the group itself as well which can be a `UserGroup`.

Reply:
OK, that makes sense. Maybe we should add a  `get_recursive_strict_subgroups` that doesn't include the group itself to clean this up, since it'd let us also skip the exclude of the current group ID here.

Reply:
Added `get_recursive_strict_subgroups` function.